0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+but be duck typed (like macros) and semantically copy/pasted into the calling scope.
+This is basically the exact feature that "`macro_rules!` with function-like captures" is trying to serve,
+except for one important thing: a `macro fn` is (potentially) still a `fn` in that it has one fixed airity,
+and can't be overloaded like a macro can.
+Basically, `macro fn` is asking for "macros 2.0,"
+which is still desirable, but still a _long_ ways off.
+`$:value` offers a small improvement in the status quo without adding a completely new system into the language.
+
+# Rationale
+[rationale]: #rationale
+
+`$:value` simplifies the authoring of `macro_rules` macros,
+as authors now no longer need to learn and remember to use the `match` trick to bind macro value arguments,
+and instead can just use the `$:value` matcher to get function-argument semantics.
+Thus, while adding to the semantics provided by the Rust compiler,
+it reduces the needed complexity to write correct `macro_rules!`.
+
+Additionally,
+it is impossible to have a repetition of expr captures that has function-argument like drop timing through use of the `match` trick alone,
+as it requires knowing the airity of the captures ahead of time to name each capture.
+`$:value` directly unlocks properly and fully variadic macros that act like function calls with respect to temporary lifetimes.
+
+# Prior art
+[prior-art]: #prior-art
+
+Kotlin's [`inline fn`](https://kotlinlang.org/docs/inline-functions.html) behave similarly to (typechecked) macros.
+Semantically, `inline fn` is inlined into the call site,
+allowing things like `return`/`break`/`continue`ing from the calling scope.
+Otherwise, `inline fn` behaves semantically like a function call.
+
+[POV-Ray's macros](http://www.povray.org/documentation/3.7.0/r3_3.html#r3_3_2_8_3)
+are also interesting for comparison.
+
+The author knows of no other languages with macro-like functionality
+that isn't just textual or lexical replacement.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Would `$:value` or `$:place` be more desirable to have in the language?
+  How much more complicated _would_ `$:place` be to implement?
+- How exactly should `$:value`(/`$:place`) be treated by `cargo expand` (`-Z unpretty`)?
+  - `$:value` can probably just use the `match` trick,
+    reducing semantic loss to just hygiene/name clashes,
+    which is already expected.
+  - `$:place` is a big unknown,
+    and probably the best it can do is a `$:value` match
+    by-value, by-ref, or by-ref-mut as appropriate.
+- Do we want to (can we) provide similar "correctness shortcuts" to proc macros,
+  rather than leaving them to rely on the `match` trick?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+If `$:value` is accepted now, `$:place` could be accpeted later if deemed necessary.
+`$:place` is usable in all the places `$:value` would be used
+(and saves a semantic move/copy before optimization,)
```

Comment:
Spurious comma.

Reply:
This comma is required for the `, but`; where should it go, I don't know at this point.